PLAT-10342 Datafeed service

symphony-youness opened the pull request:
### Ticket
[PLAT-10342](https://perzoinc.atlassian.net/browse/PLAT-10342)

### Description
Datafeed service implementation

Another ticket will be implementing 
- [ ] Unit tests updated or added
- [ ] Updated the documentation in [docs folder](../docs)

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Docstrings added or updated

symphony-youness committed:
PLAT-10342 Datafeed service
- datafeed config model
- datafeed example
- creating and persisting datafeed
- reading from datafeed
- dispatching to on_message_sent event type
- fixed typo in ApiClientFactory class

symphony-youness committed:
PLAT-10342 Modified api_client template
An empty read datafeed returns a 204 response status - empty content
We have to handle it manually

symphony-elias commented on symphony/bdk/core/datafeed/abstract_datafeed_loop.py:
```diff
@@ -0,0 +1,79 @@
+
+    def handle_v4_event_list(self, events: List[V4Event]):
+        for event in events:
+            if event.type is None:  # should we ever check this ? Isn't the generated code supposed to check for nullity or not the right type and raise an exception
```
Indeed, I don't know if we can get an empty event type from the read DF endpoint.

symphony-youness replied:
My understanding from the generated code in the `V4Event` class is that:
- whatever value the type is, if present, it's going to be returned.
- if no type is present in the response, then there no attribute `event.type` will be created. Which will raise an exception.
-> I will handle it using a try except block in the `dispatch_on_event_type` function

symphony-elias commented on symphony/bdk/core/datafeed/abstract_datafeed_loop.py:
```diff
@@ -0,0 +1,79 @@
+        except KeyError:
+            print(f"Received event with an unknown type: {event.type}")
+
+    def handle_event(self, event: V4Event):
```
I guess you can remove that

symphony-elias commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,48 @@
+            self.started = True
+            while self.started:  # I would like to know why do we use a do while in the java bdk for the read datafeed ?
+                await self.read_datafeed()
+        except ApiException:
```
You should filter on HTTP 400 and this should be done inside the while loop.

symphony-youness replied:
Please tell me if I am wrong but we don't really get a response if the status code is 400, the generated code always throws an exception. **I will look more into this and how to get the status code.**
<br>
I tried to handle it like this, but if there is a problem with the read_datafeed endpoint and not the create datafeed endpoint, we end up in an endless loop.
```python
    async def start(self):
        if self.datafeed_id == "":
            self.datafeed_id = await self.create_datafeed_and_persist()
        self.started = True
        while self.started:
            try:
                await self.read_datafeed()
            except ApiException:
                try:
                    self.datafeed_id = await self.create_datafeed_and_persist()
                except ApiException:
                    raise
```

symphony-elias replied:
You can check this way: 
```
except ApiException as e:
  if e.status == 400:
    ...
```

symphony-elias commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,48 @@
+            'session_token': self.auth_session.session_token,
+            'key_manager_token': self.auth_session.key_manager_token,
+        }
+        response = await self.datafeed_api.v4_datafeed_create_post(**params)
```
Please remove the `params` dict and put directly the params like: `self.datafeed_api.v4_datafeed_create_post(session_token=self.auth_session.session_token, key_manager_token=self.auth_session.key_manager_token)`

symphony-elias commented on symphony/bdk/core/datafeed/on_disk_datafeed_id_repository.py:
```diff
@@ -0,0 +1,80 @@
+        content = datafeed_id + "@" + agent_base_path
+        try:
+            self._get_datafeed_id_file().write_text(content, encoding='utf-8')
+        except ValueError:
```
Why `ValueError`? What about I/O error (like file system permission issues)?

symphony-youness replied:
This was for handling encoding errors but the content shouldn't be other than `utf-8`.
**I will remove it**

symphony-elias commented on symphony/bdk/core/datafeed/real_time_event_listener.py:
```diff
@@ -0,0 +1,83 @@
+from symphony.bdk.gen.agent_model.v4_symphony_elements_action import V4SymphonyElementsAction
+
+
+class RealTimeEventListener:
```
I believe this should be an ABC ?

symphony-youness replied:
The user might want to implement some listeners and not all of them right? 
This is why I didn't use the @abstractmethod decorator to force the implementation of the methods.
Why make it an ABC then ?

symphony-elias replied:
👌 I get it

symphony-youness committed:
PLAT-10342 Typo in ApiClientFactory
- Typo in ApiClientFactory
- reformat a file (add a newline)

symphony-youness committed:
PLAT-10342 added all dispatch event types
- added all event types in the dispatch dict
- made method static and fixed formatting

symphony-youri commented on symphony/bdk/core/datafeed/on_disk_datafeed_id_repository.py:
```diff
@@ -0,0 +1,80 @@
+        pass
+
+    @abstractmethod
+    def read_agent_base_path(self):
```
should we add the returned types?

symphony-youri commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,48 @@
+            if self.datafeed_id == "":
+                self.datafeed_id = await self.create_datafeed_and_persist()
+            self.started = True
+            while self.started:  # I would like to know why do we use a do while in the java bdk for the read datafeed ?
```
to set the started status the first time?

symphony-youri commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,48 @@
+    def retrieve_datafeed(self):
+        return self.datafeed_repository.read()
+
+    async def create_datafeed_and_persist(self):
```
should we mark these methods as private?

symphony-youri commented on symphony/bdk/core/datafeed/abstract_datafeed_loop.py:
```diff
@@ -0,0 +1,93 @@
+        self.bdk_config = config
+        self.api_client = datafeed_api.api_client
+        self.dispatch_dict = {
+            'MESSAGESENT': ('on_message_sent', 'message_sent'),
```
is there a way in Python we could use method reference directly instead of strings?

symphony-youri replied:
Something like https://stackoverflow.com/questions/1351669/referencing-class-methods-in-class-lists-in-python maybe?

symphony-youness replied:
I spent some time on that one, I agree with you it could have been better.

The problem with this is that we want to access the user implementation of the `on_message_sent`function. 
We can do it using `listener.on_message_sent`. But we can't really pass the listener to a dict or an enum even.

A better solution was using decorators (@singledispatch) but I found it a bit complex to implement. So I'll implement it as an improvement, once I figure out how it works.

symphony-youri commented on symphony/bdk/core/config/model/bdk_datafeed_config.py:
```diff
@@ -0,0 +1,17 @@
+class BdkDatafeedConfig:
+
+    def __init__(self, config):
+        self.id_file_path = ""
```
should we read it from the config?

symphony-youness committed:
PLAT-10342 fixed params and others
- fixed params in read and create datafeed calls
- removed ValueError handling when writing the datafeed id
- removed unused function

symphony-youness committed:
fixup! PLAT-10342 removed datafeed coverage
- removed coverage of the datafeed package, added test of bdk_datafeed_config

thibauult commented on pyproject.toml:
```diff
@@ -33,7 +33,7 @@ junit_logging = "all"
 branch = true
 source = ["symphony"]
-omit = ["symphony/bdk/gen/*"]
+omit = ["symphony/bdk/gen/*", "symphony/bdk/core/datafeed/*"]
```
We won't test this feature?

symphony-youness replied:
I will create a ticket with no points to add tests for next sprint. 

This is purely for the build to pass.

symphony-youness committed:
PLAT-10342 added "idFilePath" to config
- added "idFilePath" to datafeed configuration
- updated tests

symphony-youness committed:
PLAT-10342 updated start datafeed to recreate the datafeed id on exception
- updated start datafeed to recreate the datafeed id in the case of an error on the read_datafeed
- added type hints
- made functions private

symphony-youness committed:
PLAT-10342 Added docstrings
- Added docstrings to classes and functions
- fixed typo

symphony-youness committed:
PLAT-10342 Fixed datafeed.start() error handling

symphony-elias commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -12,34 +31,34 @@ def __init__(self, datafeed_api, auth_session, config, repository=None):
         self.started = False
 
-    async def read_datafeed(self):
+    async def __read_datafeed(self):
```
Single underscore would be enough.

symphony-elias commented on symphony/bdk/core/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,62 @@
+                await self.__read_datafeed()
+            except ApiException as e:
+                if e.status == 400:
+                    self.datafeed_id = await self.__create_datafeed_and_persist()
```
else we raise the `ApiException` because we should only filter HTTP 400 errors.

symphony-youness committed:
PLAT-10342 Fix: datafeed.start() error handling

symphony-youness committed:
PLAT-10342 updated double underscores to single ones

symphony-youness committed:
PLAT-10342 Moved datafeed package to symphony/core/service

symphony-youness committed:
Merge with upstream 2.0
- message service
- updated datafeed to use the new SymphonyBdk entry point

symphony-youness committed:
PLAT-10342 handled event type attribute error

symphony-youness committed:
PLAT-10342 Datafeed V1 tests
- added tests for the datafeed loop
- updated pyproject.toml to use python 3.8
- small fixes on the config

symphony-youness committed:
PLAT-10342 changed python version to 3.8 in circle ci config file
- 3.7 -> 3.8 on circle ci config file

thibauult commented on tests/resources/datafeed/datafeedId:
```diff
@@ -0,0 +1 @@
+8e7c8672-220@localhost:7443
```
To be ignored from `.gitignore`

symphony-youri replied:
Sure? it seems used in the tests

thibauult commented on symphony/bdk/core/service/datafeed/on_disk_datafeed_id_repository.py:
```diff
@@ -0,0 +1,78 @@
+from symphony.bdk.core.config.model.bdk_config import BdkConfig
+
+
+class DatafeedIdRepository(ABC):
```
This file is called `on_disk_datafeed_id_repository.py`, however you define the `DatafeedIdRepository` interface + the `OnDiskDatafeedIdRepository` implem. 

I'd move those classes to `symphony/bdk/core/service/datafeed/__init__.py` to make imports simpler

symphony-elias replied:
@symphony-thibault I understand you don't want to put the interface and implementation in the same module. But I don't like putting code in the `__init__.py` files.
We could split this into two modules and put import statements in the `__init__.py` files to simplify imports.

symphony-youri commented on docs/datafeed.md:
```diff
@@ -0,0 +1,21 @@
+
+
+## How to use
+The central component for the contract between bot developers and  the Datafeed API is the `DatafeedService`.
```
extra space between and..the (yes I'm a maniac)

symphony-youri replied:
DatafeedService -> Datafeed Loop?

symphony-youri commented on symphony/bdk/core/service/datafeed/abstract_datafeed_loop.py:
```diff
@@ -84,10 +88,8 @@ def dispatch_on_event_type(self, listener: RealTimeEventListener, event: V4Event
             listener_method = getattr(listener, listener_method_name)
             listener_method(event.initiator, getattr(event.payload, payload_field_name))
+        except AttributeError as e:
+            print(f"Attribute error while dispatching the {event}. {e}")
```
should not we start using logging instead of printf?

symphony-youri commented on symphony/bdk/core/service/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,64 @@
+
+    This service will be started by calling :func:`~DatafeedLoopV1.start`.
+
+    On the vert first run, the datafeed will be created and the BDK bot will listen to this datafeed to receive
```
vert-> very

symphony-youri commented on symphony/bdk/core/service/datafeed/datafeed_loop_v1.py:
```diff
@@ -0,0 +1,64 @@
+        self.datafeed_repository.write(response.id)
+        return datafeed_id
+
+    def retrieve_datafeed(self):
```
do we need to make it public?

symphony-youri commented on tests/core/config/models/test_bdk_datafeed_config.py:
```diff
@@ -0,0 +1,28 @@
+import pytest
+
+
+@pytest.fixture(params=["v1", 25, True])
```
what are 25/true used for?

symphony-youri commented on tests/core/config/models/test_bdk_datafeed_config.py:
```diff
@@ -0,0 +1,28 @@
+    return request.param
+
+
+# maybe can find an alternative to cucumber
```
I didn't understand the comment
